plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -443,6 +462,57 @@ private static <K, V> Map<K, V> getAll(Cache<K, AtomicReference<V>> cache, Itera
         return result.buildOrThrow();
     }
 
+    private static <K, V> Map<K, V> getAll(
+            Cache<K, AtomicReference<V>> cache,
+            Iterable<K> keys,
+            Function<Collection<K>, Map<K, V>> bulkLoader,
+            Predicate<V> isSufficient,
+            BiFunction<V, V, V> merger)
+    {
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        Map<K, AtomicReference<V>> toLoad = new HashMap<>();
+
+        keys.forEach(key -> {
+            // make sure the value holder is retrieved before the new values are loaded
+            // so that in case of invalidation we will not set the stale value
+            AtomicReference<V> currentValueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+            V currentValue = currentValueHolder.get();
+            if (currentValue != null && isSufficient.test(currentValue)) {
+                result.put(key, currentValue);
+            }
+            else {
+                toLoad.put(key, currentValueHolder);
+            }
+        });
+
+        if (toLoad.isEmpty()) {
+            return result.buildOrThrow();
+        }
+
+        Map<K, V> newEntries = bulkLoader.apply(toLoad.keySet());
+        newEntries.forEach((key, newValue) -> {
```

Comment:
we could iterate over `toLoad` instead to make it very explicit we return a value for all keys requested (unless it's not true?)

Reply:
I think it's true, done.